Framework/NexusGeometry/src/JSONGeometryParser.cpp

```diff
@@ -0,0 +1,556 @@
+  auto datadesc = datasetParent["dataset"];
+  auto shape = datadesc["size"];
+  auto values = datasetParent[VALUES];
+
+  std::vector<uint32_t> dims(shape.size());
+
+  size_t nValues = 1;
+  for (Json::ArrayIndex i = 0; i < shape.size(); ++i) {
+    dims[i] = shape[i].asUInt();
+    nValues *= dims[i];
+  }
+
+  data.reserve(data.size() + nValues);
+
+  auto outer = dims[0];
+  for (Json::ArrayIndex i = 0; i < outer; ++i) {
+    auto val = values[i];
+    recursiveFill<T>(val, data);
+  }
+}
+
+void extractShapeInformation(const Json::Value &shape,
+                             std::vector<uint32_t> &cylinders,
+                             std::vector<uint32_t> &faces,
+                             std::vector<Eigen::Vector3d> &vertices,
+                             std::vector<uint32_t> &windingOrder,
+                             bool &isOffGeometry) {
+  auto attributes = shape[ATTRIBUTES][0];
+  auto name = shape["name"].asString();
+  std::vector<float> verts;
+
+  const auto &children = shape[CHILDREN];
+  isOffGeometry = false;
+  if (attributes[VALUES] == NX_OFF) {
+    for (const auto &child : children) {
+      if (child[NAME] == "faces")
+        extractDatasetValues<uint32_t>(child, faces);
+      else if (child[NAME] == "vertices")
+        extractDatasetValues<float>(child, verts);
+      else if (child[NAME] == "winding_order")
+        extractDatasetValues<uint32_t>(child, windingOrder);
+    }
+
+    if (windingOrder.size() != verts.size() / 3)
+      throw std::invalid_argument("Invalid off geometry provided in json " +
+                                  name + ".");
+
+    isOffGeometry = true;
+  }
+
+  if (attributes[VALUES] == NX_CYLINDER) {
+    for (const auto &child : children) {
+      if (child[NAME] == "cylinders")
+        extractDatasetValues<uint32_t>(child, cylinders);
+      else if (child[NAME] == "vertices")
+        extractDatasetValues<float>(child, verts);
+    }
+
+    if (cylinders.size() != verts.size() / 3)
+      throw std::invalid_argument(
+          "Invalid cylindrical geometry provided in json " + name + ".");
+  }
+
+  vertices.reserve(vertices.size() + (verts.size() / 3));
+
+  for (size_t i = 0; i < verts.size(); i += 3)
+    vertices.emplace_back(Eigen::Vector3d(static_cast<double>(verts[i]),
+                                          static_cast<double>(verts[i + 1]),
+                                          static_cast<double>(verts[i + 2])));
+}
+
+bool validateShapeInformation(const bool &isOffGeometry,
+                              const std::vector<Eigen::Vector3d> &vertices,
+                              const std::vector<uint32_t> &cylinders,
+                              const std::vector<uint32_t> &faces,
+                              const std::vector<uint32_t> &windingOrder) {
+  if ((isOffGeometry &&
+       (vertices.empty() || faces.empty() || windingOrder.empty())) ||
+      (!isOffGeometry && (vertices.empty() || cylinders.empty())))
+    return false;
```

Comment:
I'm not sure you should do it in this PR, but this seems like the information you might want to log. I changed `NexusGeometryParser` recently to allow logging via the `Kernel::Logger` maybe you could do the same thing here.